### ISSUE TITLE
Further reduce the size of the parse tables

### ISSUE THREAD

**Issue by Marwes** [lalrpop/lalrpop#304]
At a whim, I looked at how much memory the generated parser were taking up in the binary and it turns out it is pretty sizeable.

Gluon's grammar currently has ~320 states, ~150 symbols and ~30 terminals, which with some quick calculations gives a resulting size of ~0.25Mb.

```
__ACTION = 320 * 30
__GOTO = 320 * 150
BYTES = (__ACTION + __GOTO) * 4 bytes = 243200
```

On its own, this is not a huge part of gluons total ~3.5 Mb size but it is not a trivial part either.

So far I can think of three improvements towards this.

The easiest solution would be to just a use a smaller representation than `i32` as the state representation. With just 328 states it would easily fit in a `i16` instead. For really small grammars even an `i8` would be enough. This should be relatively easily fixed when generating the table and simply select the smallest integer which fits all states.

Another improvement which can help (if it is possible) some cases is to avoid generating unique tables for each `pub` item. Gluon exports [two](https://github.com/gluon-lang/gluon/blob/d01783e37e32a2a3315c410b79b17025b80771a2/parser/src/grammar.lalrpop#L589-L593) entrypoints and these each get their own unique tables despite being almost the same. I *think* it is possible to generate a single parser for both of these and just generate a small initializer for each which starts the parser of in the correct state after which everything should work as normal (please correct me if I am wrong about this @nikomatsakis ).

A third possibility is to use a different, more compact representation for the tables than the current straight array. This would likely incur a performance penalty so it might not be worth it.

**Comment by nikomatsakis** [comment]
There is a lot of prior art on this. A very typical thing also is to compress the tables in some way. 

> I think it is possible to generate a single parser for both of these and just generate a small initializer for each which starts the parser of in the correct state after which everything should work as normal 

Quite possibly =) I've not given it much thought.

**Comment by Marwes** [comment]
> There is a lot of prior art on this. A very typical thing also is to compress the tables in some way.

I tried to look into some prior art but I must be searching for the wrong terms since as I didn't find anything about it :/ . Should try a bit harder I guess.

I thought about mentioning compression as well but I feel that is cheating since it just serves to hide how much memory you actually need to use the parser. (+ it does not help to reduce size when distributing the library either since that will be compressed regardless).

**Comment by ahmedcharles** [comment]
Technically, compression here could (probably does) mean a data layout in memory that is decompressed for every state.

For example, you could use variable width encoding to represent the data for a state in the action table and have two arrays, one which is indexed by state and returns the address of the corresponding sequence of bytes that will then be indexed by the `integer` parameter.

https://github.com/lalrpop/lalrpop/blob/master/lalrpop/src/lr1/codegen/parse_table.rs#L102

That would become:

```
let action = ACTION[state][integer];
```
Only the second index operator would be custom and be O(n) instead of O(1).

Note, you can sync points more frequently than once per state, but that increases memory usage again.

**Comment by Marwes** [comment]
@ahmedcharles I'd be vary about the slowdown from a custom data structure in that case (mentioned in my third point), but it is worth looking into at least.

Some more context as towards what is worth optimizing. Without #306 the binary size was 50/50 split between the tables and the reduce function. So assuming #306 gets merged it will be closer to a 33/67 split (since that PR will cut memory use of the tables in half for most if not all parsers). Further optimization on the table part might still be useful but chances are it is overshadowed by the size of `reduce`.

Unfortunately I can't think of a way to actually improve on that :/

**Comment by nikomatsakis** [comment]
@Marwes regarding shrinking the reduce function, it is indeed hard to see how we can do that much about it...

**Comment by nikomatsakis** [comment]
(Unless maybe your parser winds up with various "identity reduces" or other cases we can identify and try to consolidate.)

**Comment by ahmedcharles** [comment]
Obviously, I wouldn't recommend making that change without extensive benchmarking to understand the tradeoffs, but it is an option. But it seems like focusing on the reduce function would be more beneficial.

**Comment by Marwes** [comment]
The only thing we can perhaps do on the reduce functions is to identify which symbols aren't used by name in a reduction and for those we only generate `symbols.pop()` and skip on doing the unpacking. It probably won't help much however.

**Comment by Marwes** [comment]
One thing that could reduce things further would be to merge `__Symbol` variants which hold the same type of value. The state machine never cares about the variant itself (just what value it carries) so this should let us avoid emitting numerous `pop_` functions.

**Comment by Marwes** [comment]
The parameters in the `action*` functions are frequently unused (mainly the start and end positions). If we keep track of which parameters are actually used (both when generating the action and calling the action), it would likely reduce the size of these functions a bit.

**Comment by windelbouwman** [comment]
I would be really interested in having a shared table for multiple exported entrypoints. In rustpython we use three entrypoints, which almost triples the size of the generated parser.

**Comment by Marwes** [comment]
Shared parsers could be done today with some manual work and we could support the same scheme through some codegeneration inside LALRPOP as well to remove the boiler plate
> https://gitter.im/lalrpop/Lobby?at=5bcc4ad6ab17df2631c7c5e9

Except, which is what I just thought of, we can synthesize more terminals however we want

So if we want a shared parser we generate

pub Generated : Variant = {
    "GeneratedTerminal1" UserPub1 => Variant::UserPub1(<>),
    "GeneratedTerminal2" UserPub2 => Variant::UserPub2(<>),
    ...
};

And then we only generate a parser for Generated + some wrappers around it which starts the parser with "GeneratedTerminalX" already consumed/automatically inserted at the start and that extracts the value of Variant that we want

That is something that could be done today outside of LALRPOP as well

But you'd have to create dummy terminals and the variant yourself

**Comment by Marwes** [comment]
https://github.com/lalrpop/lalrpop/pull/523 reduces the size of the goto table significantly, might be possible to event better though due to how sparse the table is.

**Comment by osa1** [comment]
Re: sharing parse tables, what I do in [parsegen][1] is, after parsing the grammar, for each `pub` non-terminal, I add a new non-terminal with just one production, and rename the original non-production to something else. For example;

```
pub MyParserEntry : MyExpr = { ... };
```

I rename this to

```
MyParserEntry1 : MyExpr = { ... };
```

then add

```
MyParserEntry : MyExpr = { <x:MyParserEntry1> => x, };
```

and maintain a mapping from these synthetic non-terminals to their states. When generating code, for each such non-terminal, I generate a function that initializes the parser with their initial state.

Here's an example from parsegen test suite: https://github.com/osa1/parsegen/blob/f2204824717042bdf85ea7a885fcccfc42335282/tests/tests.rs#L360-L473

The generated code for `pub` non-terminals in the linked example looks like this:

```rust
pub struct AppExpr;
impl AppExpr {
    pub fn parse<E: Clone>(
        mut input: impl Iterator<Item = Result<Token, E>>,
    ) -> Result<Expr, ParseError<E>> {
        parse_generic(input, SemanticActionResult::non_terminal_0, 2u32, 8usize)
    }
}
pub struct SimpleExpr;
impl SimpleExpr {
    pub fn parse<E: Clone>(
        mut input: impl Iterator<Item = Result<Token, E>>,
    ) -> Result<Expr, ParseError<E>> {
        parse_generic(input, SemanticActionResult::non_terminal_0, 3u32, 10usize)
    }
}
pub struct BinOpExpr;
impl BinOpExpr {
    pub fn parse<E: Clone>(
        mut input: impl Iterator<Item = Result<Token, E>>,
    ) -> Result<Expr, ParseError<E>> {
        parse_generic(input, SemanticActionResult::non_terminal_0, 0u32, 2usize)
    }
}
pub struct UnOpExpr;
impl UnOpExpr {
    pub fn parse<E: Clone>(
        mut input: impl Iterator<Item = Result<Token, E>>,
    ) -> Result<Expr, ParseError<E>> {
        parse_generic(input, SemanticActionResult::non_terminal_0, 1u32, 5usize)
    }
}
```

`parse_generic` `u32` arguments are for the initial states, `usize` arguments are for the semantic action indices. The `usize` argument could be a function pointer, but for now I pass an index to the array that holds pointers to semantic action functions.

Note that I don't pass a parse table to `parse_generic` calls: they all use the same parse table.

Just an idea.. I don't know if this is the best way to implement this.

[1]: https://github.com/osa1/parsegen
